### PR TITLE
Syntactic sugar for indexed update functions.

### PR DESCRIPTION
**Pull request by danieldjohnson**
@dpfau suggested adding some syntactic sugar for `jax.ops.index_update` and related functions. This PR implements one option for that syntactic sugar: instead of
```python
x = jax.index_update(x, jax.ops.index[..., 1:3, :], y)
x = jax.index_add(x, jax.ops.index[..., 1:3, :], y)
x = jax.index_min(x, jax.ops.index[..., 1:3, :], y)
x = jax.index_max(x, jax.ops.index[..., 1:3, :], y)
```
you can write
```python
x = x.at[..., 1:3, :].set(y)
x = x.at[..., 1:3, :].add(y)
x = x.at[..., 1:3, :].min(y)
x = x.at[..., 1:3, :].max(y)
```

## Other possible options
Originally implemented by this PR:
```python
x = x.update[..., 1:3, :](y)
x = x.add[..., 1:3, :](y)  # or x = x.update_add[..., 1:3, :](y)
x = x.update_min[..., 1:3, :](y)
x = x.update_max[..., 1:3, :](y)
```
Other easy-to-implement options discussed:
```python
x = x.at[..., 1:3, :].update(y)
x = x.at[..., 1:3, :].add(y)
x = x.at[..., 1:3, :].min(y)
x = x.at[..., 1:3, :].max(y)
```

```python
x = x.update[..., 1:3, :].to(y)
x = x.update[..., 1:3, :].add(y)
x = x.update[..., 1:3, :].min(y)
x = x.update[..., 1:3, :].max(y)
```

Harder to implement (would require `x[idx]` to act both as an array itself and a slice reference to `x`):
```python
x = x[..., 1:3, :].update(y)
x = x[..., 1:3, :].update_add(y)
x = x[..., 1:3, :].update_min(y)
x = x[..., 1:3, :].update_max(y)
```

Even harder to implement (would require adding a layer of indirection)
```python
# x is an indirect wrapper of a device value,
# where __setattr__ changes the wrapped value
x[..., 1:3, :] = y
x[..., 1:3, :] += y
# (mutable) x now contains a different (immutable) wrapped value
```

I'd be interested to know which of these people prefer the best, and if there are other tradeoffs here I haven't thought of.

### CODE REVIEW

**Comment by dpfau** [comment]
Personally I like

```
x = x.at[..., 1:3, :].update(y)
x = x.at[..., 1:3, :].add(y)
x = x.at[..., 1:3, :].min(y)
x = x.at[..., 1:3, :].max(y)
```

with maybe additional sugar so that 
```
x = x.at[..., 1:3, :](y)
```
is shorthand for 
```
x = x.at[..., 1:3, :].update(y)
```
as that is the most common case. I like having the brackets and parentheses separated and having consistent naming for the function that the slice is applied to. What do other people think?

**Comment by dpfau** [comment]
One other possibility...would it make sense to do `set` instead of `update`? Then all the update functions would be equal length (`add`, `min`, `max`). Possibly overkill, but I do appreciate concision.

**Comment by shoyer** [comment]
My preference is definitely for one of the slightly longer forms like `x.at[..., 1:3, :].update(y)` or `x.update[..., 1:3, :].to(y)`, for two reasons:

1. It's much more readable. It's much less obvious to me what `x.at[..., 1:3, :](y)` would mean.
2. All of these methods end up in a single "namespace" on JAX array objects, which makes them more discoverable.

RE: the `set` name instead of `update`: I think `set` would be fine, too, as would be `assign` (which `tf.Variable` uses). `update` has two minor virtues: (1) it's consistent with the existing name for these ops in JAX and (2) there's no confusion with Python's builtin `set()` objects. If we were copying a name from builtin Python, the more natural choice might be `setitem`.

**Comment by dpfau** [comment]
So how does a decision like this get made? Is Matt benevolent dictator for life, or do you have some complicated parliamentary procedure involving terms like "cloture" and "quorum" and following Roberts Rules of Order?

**Comment by shoyer** [comment]
Ideally we reach consensus, but Matt can definitely be a tie-breaker :)

@hexahedria which syntax do you prefer?

**Comment by hexahedria** [comment]
I think I'm starting to prefer the longer namespaced forms (`x.at[idx].update(y)` and `x.update[idx].to(y)`) as well.

I don't have strong opinions about the exact words used, but here's a fairly weak one: `at` seems slightly less self-explanatory than `update`, in the sense that `x.at[idx]` isn't obviously distinct from `x[idx]` for someone who hasn't seen this syntactic sugar before. So that might be a slight point in favor of `x.update[idx]` over `x.at[idx]`. (Relatedly, if we want to allow shorthand I think `x.update[idx](y)` seems pretty clear but `x.at[idx](y)` is potentially confusing.) On the other hand `at` is shorter and reads nicely, whereas `update` is a bit verbose, so I could see it going either way.

**Comment by mattjj** [comment]
Wow, beautifully done @hexahedria ! This LGTM in general. I don't have strong opinions here and it sounds quite likely that we can find a consensus. We could even make a poll in one of our chat rooms if we really wanted...

**Comment by dpfau** [comment]
Well let me stew over it a bit more. I do like the `at` syntax a little more (again, for concision) but consistency with the unsugared syntax is probably good too. Also note that since PR #2696, there is now an `index_mul` as well, and having some sugar for pure updates to diagonals (instead of the in-place `np.fill_diagonal` I brought up in #2680) is probably a good idea too.

**Comment by dpfau** [comment]
How far away from the original NumPy interface are people ok with going?

**Comment by shoyer** [comment]
> How far away from the original NumPy interface are people ok with going?

I think we're already deviating pretty far here. I'm open to nearly anything as long as it's clearly an extension, and isn't valid NumPy code (there is a much higher bar for incompatible changes).

**Comment by mattjj** [comment]
FWIW the chat room poll was pretty decisive: +8 for `x = x.at[..., 1:3, :].add(y)` etc, and +1 for `x = x.update[..., 1:3, :].add(y)`.

**Comment by dpfau** [comment]
There was a poll!!!

What about `update` vs `set`?

**Comment by dpfau** [comment]
Thoughts on alternatives to `fill_diagonal`? `x.at[..., 1:3, :].diag(y)` ?

**Comment by mattjj** [comment]
I didn't ask it directly, but Jascha independently suggested `set`. (The poll is actually in the JAXers thread you started btw, though that chat is so high volume it's pretty hard to keep track!)

Haven't thought about `fill_diagonal`...

**Comment by shoyer** [comment]
We didn't poll `update` vs `set`.

I think `fill_diagonal` should probably just be a separate function in `jax.ops`? There isn't any syntax advantage to using a property, since there is no indexing involved. Also indexed updates/arithmetic have builtin syntax support in NumPy, so it seems more appropriate to use a method. In contrast `np.fill_diagonal` is a function.

**Comment by hexahedria** [comment]
It looks like people are converging on `x.at[idx].set(y)`, so I've switched this PR to implement that syntax instead of using my original names.

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3787,3 +3788,115 @@ def _unstack(x):
     raise ValueError("Argument to _unstack must be non-scalar")
   return [lax.index_in_dim(x, i, keepdims=False) for i in range(x.shape[0])]
 setattr(DeviceArray, "_unstack", _unstack)
+
+
+# Syntactic sugar for scatter operations.
+class IndexUpdateHelper(object):
```
Comment (shoyer) [inline jax/numpy/lax_numpy.py:3794]: Prefix this name with an underscore. Otherwise I think it will appear in the public namespace of `jax.numpy`.

Also, no need to inherit from `object` now that we're Python 3 only :)

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -3787,3 +3788,115 @@ def _unstack(x):
...
+  # Note: this docstring will appear as the docstring for the `at` property.
+  """Indexable helper object to call indexed update functions.
+
+  The `at` property is syntactic sugar for calling the indexed update functions
+  defined in :mod:`jax.ops`, and acts as a pure equivalent of in-place
+  modificatons.
+
+  In particular:
+  - :code:`x = x.at[idx].set(y)` is a pure equivalent of :code:`x[idx] = y`.
```
Comment (shoyer) [inline jax/numpy/lax_numpy.py:3803]: Another valid option is to use double ticks here, e.g.,
```
``x = x.at[idx].set(y)``
```

Or actually I guess this doesn't really need to be valid RST at all if we aren't putting it in the docs somewhere.

**Review by shoyer — COMMENTED** [review]
Would you mind writing a brief description of how these operations work, and putting it in `docs/jax.ops.rst` so users will find it? :)

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (hexahedria) [reply jax/numpy/lax_numpy.py]: True. I guess the only time someone would see this would be if they were either reading the source code or using  `help` in a REPL.

I've changed it to double ticks because that's easier to read. I suppose single ticks (or even no ticks) would also work, but I'm not sure it really matters; people reading this can probably figure it out either way.

**Review by shoyer — APPROVED** [review]

**Review by mattjj — APPROVED** [review]
Beautiful work, @hexahedria . Great implementation, great tests, great docs. Thanks so much for this contribution, and for jumping on it so quickly.

Thanks also to @dpfau for the idea, and @shoyer for the design iteration!
